src/flint/fq_nmod.jl

```diff
@@ -480,6 +480,49 @@ rand(r::Random.AbstractRNG, R::FqNmodFiniteField, b::UnitRange{Int}) = rand(r, m
 
 rand(R::FqNmodFiniteField, b::UnitRange{Int}) = rand(Random.GLOBAL_RNG, R, b)
 
+###############################################################################
+#
+#   Iteration
+#
+###############################################################################
+
+Base.iterate(F::FqNmodFiniteField) = zero(F), zeros(UInt, degree(F))
+Base.iterate(F::FqFiniteField) = zero(F), zeros(fmpz, degree(F))
+
+function Base.iterate(F::Union{FqNmodFiniteField,FqFiniteField}, coeffs::Vector)
+   deg = length(coeffs)
+   char = F isa FqNmodFiniteField ? F.p : # cheaper than calling characteristic(F)::fmpz
+                                    characteristic(F)
+   allzero = true
+   for d = 1:deg
+      if allzero
+         coeffs[d] += 1
+         if coeffs[d] == char
+            coeffs[d] = 0
+         else
+            allzero = false
+         end
+      else
+         break
+      end
+   end
+   allzero && return nothing
+
+   elt = F()
+   for d = 1:deg
+      if F isa FqNmodFiniteField
+         ccall((:nmod_poly_set_coeff_ui, libflint), Nothing,
+               (Ref{fq_nmod}, Int, UInt), elt, d - 1, coeffs[d])
```

Comment:
@tthsqe12 Is this the right way to do in flint 2.7?

Reply:
At least there is another instance of this signature (with `Ref{fq_nmod}`) in "flint/fq_nmod_embed.jl".

Reply:
It is not wrong, but it is an abuse of the fq_nmod typedef.
Iterating through fq_nmod in lexicographical order as proven to be not very useful on several occasions for me.
I am not sure it this type of thing is worth making go fast, but if it is I would be in favor of moving the state into flint after we decide what it should do. I need fq_nmod iteration in flint, and why duplicate the code in julia?

https://github.com/wbhart/flint2/blob/0932b8f1e7efc390d163c8ebbef3b8f083dda7ff/fq_nmod_mpoly/mpolyu_gcdp_zippel.c#L15

Reply:
Is this part of flint 2.7? Do you have it for `fq`?

We have applications where we need to exhaust all of the finite field, so fast is important.

Reply:
`fq_nmod_next` seems like exactly what is needed; but I didn't find `fq_next`, is there an equivalent?

> Iterating through fq_nmod in lexicographical order as proven to be not very useful on several occasions for me.

I don't really understand what you mean, when afterwards you say that you need `fq_nmod` iteration in flint...

> I would be in favor of moving the state into flint after we decide what it should do.

Maybe I misunderstand something, but the iterator's state on the julia side could simply be an `fq_nmod`, which is modified on each iteration via `fq_nmod_next` (and copied to return a distinct element at each iteration step), so nothing is new is needed besides `fq_nmod_next` (and an hypothetical `fq_next`).

Reply:
> Is this part of flint 2.7? Do you have it for fq?

It is not officially, but if we decide what it should do and make sure the version there is correct, then it could be in 2.8.

> We have applications where we need to exhaust all of the finite field, so fast is important.

Maybe fq_zech is in order?

> I don't really understand what you mean, when afterwards you say that you need fq_nmod iteration in flint...

What I was saying here is not relevant to iteration. What I meant was that if x is going is going through fq in lexigographical order, then using the values of x directly in that order might cause problems. It would have been better for me to use the values of ax+b for fixed random nonzero a, b.

Reply:
> It is not officially, but if we decide what it should do and make sure the version there is correct, then it could be in 2.8.

What I understand it does is roughy equivalent to what is needed and implemented here for iteration; i.e. without caring about the order, just give a "next" element (as fast as possible) such that all elements are visited (I guess it could go with `fq_nmod_first` or something if the order stays unspecified, i.e. doesn't necessarily start at the zero element).

> It does not exist for fq. :(

Would it be difficult to implement for Flint 2.8?

> It would have been better for me to use the values of ax+b for fixed random nonzero a, b.

Ok I see, I was also wondering how to iterate efficiently through a (randomly) shuffled list of elements. But I personally also need the "whatever order which is the fastest".

Reply:
> Maybe fq_zech is in order?

That would be great, but it is not wrapped yet :-(

Since we won't be able to use native flint function for this functionality before 2.8 is released, I would suggest we try go this PR here merged.